lib/libc/stdlib/malloc/mrs/mrs.c

```diff
@@ -935,21 +939,26 @@ print_cheri_revoke_stats(char *what, struct cheri_revoke_syscall_info *crsi,
 }
 #endif /* PRINT_CAPREVOKE */
 
-static inline void
-slab_free_next_descriptor(struct mrs_descriptor_slab *s)
+static inline void *
+slab_pop_descriptor(struct mrs_descriptor_slab *s)
 {
-	int i = s->next_descriptor++;
+	if (s->next_descriptor == s->num_descriptors)
+		return NULL;
 
-	assert(s->next_descriptor <= s->num_descriptors);
+	return (s->slab[s->next_descriptor++].ptr);
+}
 
+static inline void
+descriptor_free(void *ptr)
+{
 	/*
 	 * Doesn't matter if the underlying size isn't a 16-byte multiple
 	 * because all allocations will be 16-byte aligned.
 	 */
-	size_t len = __builtin_align_up(cheri_getlen(s->slab[i].ptr),
+	size_t len = __builtin_align_up(cheri_getlen(ptr),
```

Comment:
This isn't really related to the diff itself, but the descriptor arenas spend 16 bytes per entry to store the size of the memory region, and that seems redundant is it's available from the capability. Do we do that because `cheri_getlen()` is relatively expensive? If so, I'd expect we want to use the saved value here rather than extracting it from the capability again.

Reply:
Indeed, we don't use the .size member after we store to it. I wonder if this dates to an older MRS design point.